src/Orleans.Core/Providers/LegacyProviderConfigurator.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Orleans.LogConsistency;
+using Orleans.Runtime.Configuration;
+using Orleans.Storage;
+using Orleans.Streams;
+using System.Collections.Generic;
+using System.Linq;
+using Orleans.Runtime;
+
+namespace Orleans.Providers
+{
+    public class LegacyProviderConfigurator
+    {
+        public const string InitStageName = "ProviderInitStage";
+        public const string StartStageName = "ProviderStartStage";
+        internal const string SchedulingContextName = "Fallback";
+        internal delegate Task QueueTask(Func<Task> taskFunc);
+    }
+
+    internal static class LegacyProviderConfigurator<TLifeCycle>
+        where TLifeCycle : ILifecycleObservable
+    {
+        /// <summary>
+        /// Legacy way to configure providers. Will need to move to a legacy package in the future
+        /// </summary>
+        /// <returns></returns>
+        internal static void ConfigureServices(IDictionary<string, ProviderCategoryConfiguration> providerConfigurations, IServiceCollection services, int defaultInitStage, int defaultStartStage)
+        {
+            // if already added providers or nothing to add, skipp
+            if (services.Any(s => s.ServiceType == typeof(ProviderTypeLookup))
+                || providerConfigurations.Count == 0)
+                return;
+
+            services.AddSingleton<ProviderTypeLookup>();
+
+            foreach (var providerGroup in providerConfigurations.GroupBy(p => p.Key))
+            {
+                if (providerGroup.Key == ProviderCategoryConfiguration.STREAM_PROVIDER_CATEGORY_NAME)
+                {
+                    foreach (IProviderConfiguration providerConfig in providerGroup.SelectMany(kvp => kvp.Value.Providers.Values))
+                    {
+                        RegisterProvider<IStreamProvider>(providerConfig, services, defaultInitStage, defaultStartStage);
+                    }
+                }
+                else if (providerGroup.Key == ProviderCategoryConfiguration.STORAGE_PROVIDER_CATEGORY_NAME)
+                {
+                    services.AddSingleton<IStorageProvider>(sp => sp.GetServiceByName<IStorageProvider>(ProviderConstants.DEFAULT_STORAGE_PROVIDER_NAME));
```

Comment:
This is just creating a mapping in DI to the default provider, so that if someone resolves an IStrorageProvider (not by name) they will get the default provider..  The default provider must still be in the list of configured providers, and be registered with RegisterProvider.

Reply:
>This is just creating a mapping in DI to the default provider, so that if someone resolves an IStrorageProvider (not by name) they will get the default provider.

why do we support this? if someone resolve IStorageProvider(not by name) then orleans return the default one, this seems arbitrary. Is it a replacement for some legacy feature we support before or?

Reply:
This is convention only, and is based off of the rules around storage provider attribution.  If no provider name is indicated in the storage provider attribute, the default storage provider is assumed.  Applying this convention to DI and Names services, it means that if one asks for a storage provider by name (using getServiceByName) they get the specific service, but if they just resolve a storage provider they should get the default storage provider.  Without this, callers who just want to use the default provider would have to know the name assigned to the default provider.